Source/Scene/ModelExperimental/ModelExperimentalArticulation.js

```diff
@@ -0,0 +1,215 @@
+      return this._runtimeStages;
+    },
+  },
+
+  /**
+   * The runtime nodes that are affected by this articulation.
+   *
+   * @memberof ModelExperimentalArticulation.prototype
+   * @type {ModelExperimentalNode[]}
+   * @readonly
+   *
+   * @private
+   */
+  runtimeNodes: {
+    get: function () {
+      return this._runtimeNodes;
+    },
+  },
+});
+
+function initialize(runtimeArticulation) {
+  const articulation = runtimeArticulation.articulation;
+
+  const stages = articulation.stages;
+  const length = stages.length;
+
+  const runtimeStages = runtimeArticulation._runtimeStages;
+  const runtimeStagesByName = runtimeArticulation._runtimeStagesByName;
+  for (let i = 0; i < length; i++) {
+    const stage = stages[i];
+    const runtimeStage = new ModelExperimentalArticulationStage({
+      stage: stage,
+      runtimeArticulation: runtimeArticulation,
+    });
+
+    // Store the stages in an array to preserve the order in which
+    // they appeared in the 3D model.
+    runtimeStages.push(runtimeStage);
+
+    // Store the stages in a dictionary for retrieval by name.
+    const stageName = stage.name;
+    runtimeStagesByName[stageName] = runtimeStage;
+  }
+}
+
+/**
+ * Sets the current value of an articulation stage.
+ *
+ * @param {String} stageName The name of the articulation stage.
+ * @param {Number} value The numeric value of this stage of the articulation.
+ *
+ * @private
+ */
+ModelExperimentalArticulation.prototype.setArticulationStage = function (
+  stageName,
+  value
+) {
+  const stage = this._runtimeStagesByName[stageName];
+  if (defined(stage)) {
+    stage.currentValue = value;
+  }
+};
+
+const scratchArticulationMatrix = new Matrix4();
+const scratchNodeMatrix = new Matrix4();
+
+/**
+ * Applies the chain of articulation stages to the transform of each node that
+ * participates in the articulation. This only recomputes the node transforms
+ * if any stage in the articulation has been modified.
+ * <p>
+ * Note that this will overwrite any existing transformations on participating
+ * nodes.
+ * </p>
+ *
+ * @private
+ */
+ModelExperimentalArticulation.prototype.apply = function () {
+  if (!this._dirty) {
+    return;
+  }
+  this._dirty = false;
+
+  let articulationMatrix = Matrix4.clone(
+    Matrix4.IDENTITY,
+    scratchArticulationMatrix
+  );
+
+  let i;
+  const stages = this._runtimeStages;
+  const stagesLength = stages.length;
+
+  // Compute the result of the articulation stages...
+  for (i = 0; i < stagesLength; i++) {
+    const stage = stages[i];
+    articulationMatrix = stage.applyStageToMatrix(articulationMatrix);
+  }
+
+  // ...then apply it to the transforms of the affected nodes.
+  const nodes = this._runtimeNodes;
+  const nodesLength = nodes.length;
+  for (i = 0; i < nodesLength; i++) {
+    const node = nodes[i];
+    const transform = Matrix4.multiplyTransformation(
+      node.originalTransform,
+      articulationMatrix,
+      scratchNodeMatrix
+    );
```

Comment:
Just checking, is this the correct multiplication order? The spec says:

>The order in which the stages appear is significant. The stages' requested transformations are applied to the node in order of appearance, after the glTF node's own transformations.

Which makes me think this would be `transform = articulationMatrix * originalTransform`, the reverse of what you have here. What does the old `Model` do?

Reply:
I also got confused by this... but my order of multiplication is what old `Model` does. Everything that is passed into `applyArticulationStageMatrix` is pre-multiplied by the transformation matrix derived from the articulation stages.